util/webhook/validator_test.go

```diff
@@ -0,0 +1,265 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"testing"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+func TestValidateBoolPtrImmutable(t *testing.T) {
+	g := NewWithT(t)
```

Comment:
I tend to put this inside the subtest so the `t` it's passed is scoped to only that subtest. That way when the test fails, the error that shows up is a bit more clear since it includes the subtest's name. Gomega gives a message alluding to this and stops execution before getting to the rest of the tests.

e.g. If I make some change to force some tests to fail:
```
% go test ./util/webhook
--- FAIL: TestValidateBoolPtrImmutable (0.00s)
    validator_test.go:81:
        Unexpected error:
            ...
    --- FAIL: TestValidateBoolPtrImmutable/nil (0.00s)
        testing.go:1343: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
FAIL
FAIL    sigs.k8s.io/cluster-api-provider-azure/util/webhook     0.025s
FAIL
```
And then after I move the `g := NewWithT(t)`:
```
% go test ./util/webhook
--- FAIL: TestValidateBoolPtrImmutable (0.00s)
    --- FAIL: TestValidateBoolPtrImmutable/nil (0.00s)
        validator_test.go:81:
            Unexpected error:
                <*field.Error | 0xc00044ff40>: {
                    Type: "FieldValueInvalid",
                    Field: "Spec.Foo",
                    BadValue: <*bool | 0x0>nil,
                    Detail: "field is immutable, unable to assign a value if it was already empty",
                }
                Spec.Foo: Invalid value: "null": field is immutable, unable to assign a value if it was already empty
            occurred
    --- FAIL: TestValidateBoolPtrImmutable/no_change (0.00s)
        validator_test.go:81:
            Unexpected error:
                <*field.Error | 0xc0002fa040>: {
                    Type: "FieldValueInvalid",
                    Field: "Spec.Foo",
                    BadValue: <*bool | 0x0>nil,
                    Detail: "field is immutable, unable to assign a value if it was already empty",
                }
                Spec.Foo: Invalid value: "null": field is immutable, unable to assign a value if it was already empty
            occurred
    --- FAIL: TestValidateBoolPtrImmutable/can't_set_from_empty (0.00s)
        validator_test.go:76:
            Expected an error to have occurred.  Got:
                <*field.Error | 0x0>: nil
FAIL
FAIL    sigs.k8s.io/cluster-api-provider-azure/util/webhook     0.023s
FAIL
```

Reply:
Thanks! updated